File: android/app/src/main/kotlin/com/transport_system/ts_driver/clock_in_out_location/FirebaseLocationLogUpdater.kt
package com.transport_system.ts_driver.clock_in_out_location

import android.content.Context
import android.location.Location
import android.util.Log
import com.google.firebase.database.FirebaseDatabase
import com.transport_system.ts_driver.data_providers.MyDetails
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class FirebaseLocationLogUpdater {

    fun updateLocationInFirebase(context: Context, location : Location){
        val myDetails : MyDetails = MyDetails.loadFromSharedPrefs(context) ?: return
        val clockInOutSession : ClockInOutSession = ClockInOutSession.loadFromSharedPrefs(context)
            ?: return
        val dataToUpdate = mapOf(
            "user_id" to myDetails.userId,
            "timestamp" to mapOf(".sv" to "timestamp"),
            "name" to "${myDetails.firstName} ${myDetails.lastName}",
            "lat" to location.latitude,
            "lng" to location.longitude,
        )
        updateLocationLog(myDetails,clockInOutSession, dataToUpdate)
        updateLiveLocation(myDetails,clockInOutSession, dataToUpdate)

    }

    // function to update live location in the Firebase
    @OptIn(DelicateCoroutinesApi::class)
    private fun updateLiveLocation(myDetails: MyDetails, clockInOutSession: ClockInOutSession, data : Map<String,Any?>){
        val root = if (clockInOutSession.isStagingServer == true)
            "user_tracking_live_Staging"
        else "user_tracking_live"

        GlobalScope.launch(Dispatchers.IO) {
            val liveRef = FirebaseDatabase.getInstance()
                .reference
                .child(root)
                .child(getCurrentDateInFormat())
                .child(myDetails.userId!!.toString())

            // updating live location
            try {

                liveRef.get().addOnCompleteListener {
                    if(it.isSuccessful){
                        val snapShot = it.result

                        if (snapShot.exists() && snapShot.value is Map<*, *>) {
                            val snapValue = snapShot.value as Map<*, *>
                            val firstKey = snapValue.keys.first()

                            // Document exists, update its data
                            if (firstKey == (clockInOutSession.sessionId!!)) {
                                liveRef.child(clockInOutSession.sessionId!!).updateChildren(data)
                            } else {
                                // Document exists, but with old collection, remove and create new
                                liveRef.removeValue()
                                liveRef.child(clockInOutSession.sessionId!!).setValue(data)
                            }
                        } else {
                            // If the document does not exist, create it
                            liveRef.child(clockInOutSession.sessionId!!).setValue(data)
                        }

                    }
                    else{
                        // If the document does not exist, create it
                        liveRef.child(clockInOutSession.sessionId!!).setValue(data)
                        Log.d("hashim","task not successful for live snapshot")
                    }
                }



            } catch (error:Exception) {
                Log.d("hashim",error.toString())
            }
        }


    }

    // function to update location log in the Firebase
    @OptIn(DelicateCoroutinesApi::class)
    private fun updateLocationLog(myDetails: MyDetails, clockInOutSession: ClockInOutSession, data : Map<String,Any?>){
        val root = if (clockInOutSession.isStagingServer == true)
         "user_tracking_logs_Staging"
        else "user_tracking_logs"

        GlobalScope.launch(Dispatchers.IO) {
            val logsRef = FirebaseDatabase.getInstance()
                .reference
                .child(root)
                .child(getCurrentDateInFormat())
                .child(myDetails.userId!!.toString())
                .child(clockInOutSession.sessionId!!)

            // updating log
            try {
                logsRef.push().setValue(data)
            } catch (error:Exception) {
                Log.d("hashim",error.toString())
            }
        }
    }

    private fun getCurrentDateInFormat(): String {
        val currentDate = LocalDate.now()
        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        return currentDate.format(formatter)
    }
}

File: ios/Runner/Clock In Out Location/FirebaseLocationLogUpdater.swift
//
//  FirebaseLocationLogUpdater.swift
//  Runner
//
//  Created by Hashim Khan on 14/11/2024.
//


import Foundation
import CoreLocation
import FirebaseDatabase


class FirebaseLocationLogUpdater {

    func updateLocationInFirebase( location: CLLocation,  fromGeofence:Bool = false) {
        guard let myDetails = MyDetails.loadFromSharedPrefs(),
              let clockInOutSession = ClockInOutSession.loadFromSharedPrefs() else {
            return
        }
        
        let firstName :String = myDetails.firstName ?? ""
        let lastName :String = myDetails.lastName ?? ""

        let dataToUpdate: [String: Any] = [
            "user_id": myDetails.userId ?? "",
            "timestamp": [".sv": "timestamp"],
            "name": "\(firstName) \(lastName)",
            "lat": location.coordinate.latitude,
            "lng": location.coordinate.longitude,
            "fromGeofence" : fromGeofence
        ]
        
        

        updateLocationLog(myDetails: myDetails, clockInOutSession: clockInOutSession, data: dataToUpdate)
        updateLiveLocation(myDetails: myDetails, clockInOutSession: clockInOutSession, data: dataToUpdate)
    }

    private func updateLiveLocation(myDetails: MyDetails, clockInOutSession: ClockInOutSession, data: [String: Any]) {
        let root = clockInOutSession.isStagingServer == true ? "user_tracking_live_Staging" : "user_tracking_live"

        let liveRef = Database.database().reference().child(root)
            .child(getCurrentDateInFormat())
            .child("\(myDetails.userId ?? 0)")

        liveRef.getData(completion: { error ,snapshot  in
            if ((snapshot?.exists()) != nil), let snapValue = snapshot?.value as? [String: Any], let firstKey = snapValue.keys.first {
                if firstKey == clockInOutSession.sessionId {
                    liveRef.child(clockInOutSession.sessionId ?? "session_error_ios").updateChildValues(data)
                } else {
                    liveRef.removeValue()
                    liveRef.child(clockInOutSession.sessionId ?? "session_error_ios").setValue(data)
                }
            } else {
                liveRef.child(clockInOutSession.sessionId ?? "session_error_ios").setValue(data)
            }
        })
    }

    private func updateLocationLog(myDetails: MyDetails, clockInOutSession: ClockInOutSession, data: [String: Any]) {
        let root = clockInOutSession.isStagingServer == true ? "user_tracking_logs_Staging" : "user_tracking_logs"

        let logsRef = Database.database().reference().child(root)
            .child(getCurrentDateInFormat())
            .child("\(myDetails.userId ?? 0)")
            .child(clockInOutSession.sessionId ?? "")

        logsRef.childByAutoId().setValue(data)
    }

    private func getCurrentDateInFormat() -> String {
        let currentDate = Date()
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: currentDate)
    }
}
